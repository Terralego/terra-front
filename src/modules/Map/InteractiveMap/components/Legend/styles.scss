@import "node_modules/@blueprintjs/core/lib/scss/variables";

.tf-legend {
  box-shadow: 0 1px 3px 0 rgba(16, 22, 26, 0.40);
  border-radius: 2px;
  background: $pt-app-background-color;
  padding: .5rem 0;
  margin-top: 1rem;
  display: flex;
  flex-direction: column;
  min-width: 250px;

  .bp3-dark & {
    background: $pt-dark-app-background-color;
  }

  > .tf-legend {
    border: 0;
  }

  &__title {
    margin: 0.2rem 0.3rem;
    font-size: 1rem;
    font-weight: 600;
    text-align: center;
  }

  &__list:not(.tf-legend__stacked-circles) {
    display: table;
    margin: 0.5rem;
    position: relative;
    min-height: 160px;

    .tf-legend__item--circle {
      display: table-row;
    }

    .tf-legend__item--square {
      padding-top: 0.5rem;

      .tf-legend__label {
        padding-bottom: 4px;
      }
    }
  }

  &__stacked-circles {
    display: table;
    margin: 1rem .5rem .5rem;
    position: relative;

    .tf-legend__item--circle {
      position: absolute;
      bottom: 0;

      .tf-legend__symbol-container {
        line-height: 0;

        &:after {
          content: "";
          width: 70%;
          border-bottom: 1px solid #000000;
          height: 1px;
          display: block;
          position: absolute;
          top: 0;
          left: 50%;
        }

        :last-child {
          &:after {
            top: 6px;
          }
        }

        .tf-legend__symbol {
          position: relative;

          svg {
            border: 1px solid #000000;
            border-radius: 50%;
          }
        }
      }

      .tf-legend__label {
        position: absolute;
        top: -8px;
        left: 120%;
        white-space: nowrap;
      }
    }
  }

  &__symbol-container {
    display: table-cell;
    text-align: center;
    vertical-align: middle;
  }

  &__symbol {
    border-radius: 0;
    display: inline-block;
  }

  &__label {
    font-size: 0.8rem;
    padding: 0 0 0 .5rem;
    display: table-cell;
    max-width: 170px;
    vertical-align: middle;
  }

  &__source {
    padding: .5em .5em 0 .5em;
    opacity: .5;
  }
}
