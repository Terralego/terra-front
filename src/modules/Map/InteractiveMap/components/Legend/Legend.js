import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import './styles.scss';

import Template from '../../../../Template/Template';
import CustomComponents from './CustomComponents';
import Circle from './components/Circle';
import Rect from './components/Rect';

const DEFAULT_DIAMETER = 16;

export const Legend = ({
  title,
  source,
  items,
  level,
  position,
  content,
  history,
  stackedCircles,
}) => {
  const biggestDiameter = items.reduce((int, item) =>
    Math.max(int, item.diameter || DEFAULT_DIAMETER), 0);

  return (
    <div className={`tf-legend tf-legend--level-${level}`}>
      <h4
        className="tf-legend__title"
      >
        {title}
      </h4>
      <div
        className={classnames(
          'tf-legend__list',
          { 'tf-legend__stacked-circles': stackedCircles },
        )}
        // The width is multiplied by a factor 1.7 in order to provide space for the legend labels
        style={stackedCircles ? { height: biggestDiameter, width: biggestDiameter * 1.7 } : null}
      >
        {content && (
          <Template
            template={content}
            customComponents={CustomComponents}
            history={history}
          />
        )}
        {items.map(({
          label,
          template,
          source: subSource,
          color,
          items: subItems,
          shape = 'square',
          radius,
          diameter = radius || DEFAULT_DIAMETER,
        }, index) => {
          if (subItems) {
            return (
              <Legend
                key={`${label}${items.length}`}
                title={label}
                items={subItems}
                source={subSource}
                level={level + 1}
              />
            );
          }

          if (template) {
            return (
              <Template
                key={index} // eslint-disable-line react/no-array-index-key
                template={template}
                customComponents={CustomComponents}
                history={history}
              />
            );
          }

          return (
            <div
              key={`${label}${color}`}
              className={`tf-legend__item tf-legend__item--${shape} tf-legend__item--${position}`}
            >
              <div
                className="tf-legend__symbol-container"
                style={{ width: shape === 'circle' ? biggestDiameter : DEFAULT_DIAMETER * 2 }}
              >
                {shape === 'circle'
                  ? <Circle color={color} size={diameter} />
                  : <Rect color={color} size={DEFAULT_DIAMETER} />}
              </div>
              <div className="tf-legend__label">{label}</div>
            </div>
          );
        })}
      </div>
      {source && <div className="tf-legend__source">{source}</div>}
    </div>
  );
};

Legend.propTypes = {
  title: PropTypes.string.isRequired,
  items: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    source: PropTypes.string,
    color: PropTypes.string,
    items: PropTypes.array,
    shape: PropTypes.string,
    diameter: PropTypes.number,
    template: PropTypes.string,
  })),
  content: PropTypes.string,
  position: PropTypes.string,
  level: PropTypes.number,
  stackedCircles: PropTypes.bool,
};
Legend.defaultProps = {
  position: 'left',
  level: 1,
  items: [],
  content: '',
  stackedCircles: false,
};

export default Legend;
