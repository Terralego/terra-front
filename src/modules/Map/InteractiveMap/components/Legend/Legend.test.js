import React from 'react';
import renderer from 'react-test-renderer';

import Legend from './Legend';

describe('should render correctly', () => {
  const items = shape => ([{
    label: 'Rouge',
    color: 'red',
    shape,
  }, {
    label: 'Vert',
    color: 'green',
    shape,
  }, {
    label: 'Bleu',
    color: 'blue',
    shape,
  }]);
  it('square', () => {
    const tree = renderer.create(<Legend title="Hello World" items={items()} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('circle', () => {
    const tree = renderer.create(<Legend title="Hello World" items={items('circle')} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('stacked circles', () => {
    const tree = renderer.create(<Legend title="Hello World" stackedCircles items={items('circle')} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('with multi level', () => {
    const tree = renderer.create((
      <Legend
        title="level1"
        items={[{
          label: 'level1.1',
          color: 'red',
        }, {
          label: 'level1.2',
          color: 'blue',
        }, {
          template: 'custom <strong>nunjucks</strong> template',
        }, {
          label: 'level2',
          items: [{
            label: 'level2.1',
            color: 'red',
          }, {
            label: 'level2.2',
            color: 'blue',
          }, {
            label: 'level3',
            items: [{
              label: 'level3.1',
              color: 'red',
            }, {
              label: 'level3.2',
              color: 'blue',
            }, {
              label: 'level4',
              items: [{
                label: 'level4.1',
                color: 'red',
              }, {
                label: 'level4.2',
                color: 'blue',
              }],
            }],
          }],
        }]}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('with source', () => {
    const tree = renderer.create(<Legend source="Evian" title="les bains" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

it('should render a manual legend', () => {
  const template =
`# Hello World

<square color="red" /> foo  
<square color="green" /> bar  

<circle color="red" /> foo  
<circle color="green" /> bar  
`;
  const tree = renderer.create(
    <Legend
      title="foo"
      content={template}
    />,
  );
  expect(tree.toJSON()).toMatchSnapshot();
});
