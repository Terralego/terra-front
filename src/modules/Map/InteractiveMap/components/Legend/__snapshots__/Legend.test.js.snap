// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a manual legend 1`] = `
<div
  className="tf-legend tf-legend--level-1"
>
  <h4
    className="tf-legend__title"
  >
    foo
  </h4>
  <div
    className="tf-legend__list"
    style={null}
  >
    <h1
      id="hello-world"
    >
      Hello World
    </h1>
    

    <p>
      <div
        className="tf-legend__symbol"
      >
        <svg
          height={10}
          viewBox="0 0 100 100"
          width={10}
        >
          <rect
            fill="red"
            height="100"
            width="100"
            x="0"
            y="0"
          />
        </svg>
      </div>
       foo
      <br />
      <div
        className="tf-legend__symbol"
      >
        <svg
          height={10}
          viewBox="0 0 100 100"
          width={10}
        >
          <rect
            fill="green"
            height="100"
            width="100"
            x="0"
            y="0"
          />
        </svg>
      </div>
       bar  
    </p>
    

    <p>
      <div
        className="tf-legend__symbol"
      >
        <svg
          height={10}
          viewBox="0 0 100 100"
          width={10}
        >
          <circle
            cx="50"
            cy="50"
            fill="red"
            r="50"
          />
        </svg>
      </div>
       foo
      <br />
      <div
        className="tf-legend__symbol"
      >
        <svg
          height={10}
          viewBox="0 0 100 100"
          width={10}
        >
          <circle
            cx="50"
            cy="50"
            fill="green"
            r="50"
          />
        </svg>
      </div>
       bar  
    </p>
    

  </div>
</div>
`;

exports[`should render correctly circle 1`] = `
<div
  className="tf-legend tf-legend--level-1"
>
  <h4
    className="tf-legend__title"
  >
    Hello World
  </h4>
  <div
    className="tf-legend__list"
    style={null}
  >
    
    <div
      className="tf-legend__item tf-legend__item--circle tf-legend__item--left"
    >
      <div
        className="tf-legend__symbol-container"
        style={
          Object {
            "width": 16,
          }
        }
      >
        <div
          className="tf-legend__symbol"
        >
          <svg
            height={16}
            viewBox="0 0 100 100"
            width={16}
          >
            <circle
              cx="50"
              cy="50"
              fill="red"
              r="50"
            />
          </svg>
        </div>
      </div>
      <div
        className="tf-legend__label"
      >
        Rouge
      </div>
    </div>
    <div
      className="tf-legend__item tf-legend__item--circle tf-legend__item--left"
    >
      <div
        className="tf-legend__symbol-container"
        style={
          Object {
            "width": 16,
          }
        }
      >
        <div
          className="tf-legend__symbol"
        >
          <svg
            height={16}
            viewBox="0 0 100 100"
            width={16}
          >
            <circle
              cx="50"
              cy="50"
              fill="green"
              r="50"
            />
          </svg>
        </div>
      </div>
      <div
        className="tf-legend__label"
      >
        Vert
      </div>
    </div>
    <div
      className="tf-legend__item tf-legend__item--circle tf-legend__item--left"
    >
      <div
        className="tf-legend__symbol-container"
        style={
          Object {
            "width": 16,
          }
        }
      >
        <div
          className="tf-legend__symbol"
        >
          <svg
            height={16}
            viewBox="0 0 100 100"
            width={16}
          >
            <circle
              cx="50"
              cy="50"
              fill="blue"
              r="50"
            />
          </svg>
        </div>
      </div>
      <div
        className="tf-legend__label"
      >
        Bleu
      </div>
    </div>
  </div>
</div>
`;

exports[`should render correctly square 1`] = `
<div
  className="tf-legend tf-legend--level-1"
>
  <h4
    className="tf-legend__title"
  >
    Hello World
  </h4>
  <div
    className="tf-legend__list"
    style={null}
  >
    
    <div
      className="tf-legend__item tf-legend__item--square tf-legend__item--left"
    >
      <div
        className="tf-legend__symbol-container"
        style={
          Object {
            "width": 32,
          }
        }
      >
        <div
          className="tf-legend__symbol"
        >
          <svg
            height={16}
            viewBox="0 0 100 100"
            width={16}
          >
            <rect
              fill="red"
              height="100"
              width="100"
              x="0"
              y="0"
            />
          </svg>
        </div>
      </div>
      <div
        className="tf-legend__label"
      >
        Rouge
      </div>
    </div>
    <div
      className="tf-legend__item tf-legend__item--square tf-legend__item--left"
    >
      <div
        className="tf-legend__symbol-container"
        style={
          Object {
            "width": 32,
          }
        }
      >
        <div
          className="tf-legend__symbol"
        >
          <svg
            height={16}
            viewBox="0 0 100 100"
            width={16}
          >
            <rect
              fill="green"
              height="100"
              width="100"
              x="0"
              y="0"
            />
          </svg>
        </div>
      </div>
      <div
        className="tf-legend__label"
      >
        Vert
      </div>
    </div>
    <div
      className="tf-legend__item tf-legend__item--square tf-legend__item--left"
    >
      <div
        className="tf-legend__symbol-container"
        style={
          Object {
            "width": 32,
          }
        }
      >
        <div
          className="tf-legend__symbol"
        >
          <svg
            height={16}
            viewBox="0 0 100 100"
            width={16}
          >
            <rect
              fill="blue"
              height="100"
              width="100"
              x="0"
              y="0"
            />
          </svg>
        </div>
      </div>
      <div
        className="tf-legend__label"
      >
        Bleu
      </div>
    </div>
  </div>
</div>
`;

exports[`should render correctly stacked circles 1`] = `
<div
  className="tf-legend tf-legend--level-1"
>
  <h4
    className="tf-legend__title"
  >
    Hello World
  </h4>
  <div
    className="tf-legend__list tf-legend__stacked-circles"
    style={
      Object {
        "height": 16,
        "width": 27.2,
      }
    }
  >
    
    <div
      className="tf-legend__item tf-legend__item--circle tf-legend__item--left"
    >
      <div
        className="tf-legend__symbol-container"
        style={
          Object {
            "width": 16,
          }
        }
      >
        <div
          className="tf-legend__symbol"
        >
          <svg
            height={16}
            viewBox="0 0 100 100"
            width={16}
          >
            <circle
              cx="50"
              cy="50"
              fill="red"
              r="50"
            />
          </svg>
        </div>
      </div>
      <div
        className="tf-legend__label"
      >
        Rouge
      </div>
    </div>
    <div
      className="tf-legend__item tf-legend__item--circle tf-legend__item--left"
    >
      <div
        className="tf-legend__symbol-container"
        style={
          Object {
            "width": 16,
          }
        }
      >
        <div
          className="tf-legend__symbol"
        >
          <svg
            height={16}
            viewBox="0 0 100 100"
            width={16}
          >
            <circle
              cx="50"
              cy="50"
              fill="green"
              r="50"
            />
          </svg>
        </div>
      </div>
      <div
        className="tf-legend__label"
      >
        Vert
      </div>
    </div>
    <div
      className="tf-legend__item tf-legend__item--circle tf-legend__item--left"
    >
      <div
        className="tf-legend__symbol-container"
        style={
          Object {
            "width": 16,
          }
        }
      >
        <div
          className="tf-legend__symbol"
        >
          <svg
            height={16}
            viewBox="0 0 100 100"
            width={16}
          >
            <circle
              cx="50"
              cy="50"
              fill="blue"
              r="50"
            />
          </svg>
        </div>
      </div>
      <div
        className="tf-legend__label"
      >
        Bleu
      </div>
    </div>
  </div>
</div>
`;

exports[`should render correctly with multi level 1`] = `
<div
  className="tf-legend tf-legend--level-1"
>
  <h4
    className="tf-legend__title"
  >
    level1
  </h4>
  <div
    className="tf-legend__list"
    style={null}
  >
    
    <div
      className="tf-legend__item tf-legend__item--square tf-legend__item--left"
    >
      <div
        className="tf-legend__symbol-container"
        style={
          Object {
            "width": 32,
          }
        }
      >
        <div
          className="tf-legend__symbol"
        >
          <svg
            height={16}
            viewBox="0 0 100 100"
            width={16}
          >
            <rect
              fill="red"
              height="100"
              width="100"
              x="0"
              y="0"
            />
          </svg>
        </div>
      </div>
      <div
        className="tf-legend__label"
      >
        level1.1
      </div>
    </div>
    <div
      className="tf-legend__item tf-legend__item--square tf-legend__item--left"
    >
      <div
        className="tf-legend__symbol-container"
        style={
          Object {
            "width": 32,
          }
        }
      >
        <div
          className="tf-legend__symbol"
        >
          <svg
            height={16}
            viewBox="0 0 100 100"
            width={16}
          >
            <rect
              fill="blue"
              height="100"
              width="100"
              x="0"
              y="0"
            />
          </svg>
        </div>
      </div>
      <div
        className="tf-legend__label"
      >
        level1.2
      </div>
    </div>
    <p>
      custom 
      <strong>
        nunjucks
      </strong>
       template
    </p>
    

    <div
      className="tf-legend tf-legend--level-2"
    >
      <h4
        className="tf-legend__title"
      >
        level2
      </h4>
      <div
        className="tf-legend__list"
        style={null}
      >
        
        <div
          className="tf-legend__item tf-legend__item--square tf-legend__item--left"
        >
          <div
            className="tf-legend__symbol-container"
            style={
              Object {
                "width": 32,
              }
            }
          >
            <div
              className="tf-legend__symbol"
            >
              <svg
                height={16}
                viewBox="0 0 100 100"
                width={16}
              >
                <rect
                  fill="red"
                  height="100"
                  width="100"
                  x="0"
                  y="0"
                />
              </svg>
            </div>
          </div>
          <div
            className="tf-legend__label"
          >
            level2.1
          </div>
        </div>
        <div
          className="tf-legend__item tf-legend__item--square tf-legend__item--left"
        >
          <div
            className="tf-legend__symbol-container"
            style={
              Object {
                "width": 32,
              }
            }
          >
            <div
              className="tf-legend__symbol"
            >
              <svg
                height={16}
                viewBox="0 0 100 100"
                width={16}
              >
                <rect
                  fill="blue"
                  height="100"
                  width="100"
                  x="0"
                  y="0"
                />
              </svg>
            </div>
          </div>
          <div
            className="tf-legend__label"
          >
            level2.2
          </div>
        </div>
        <div
          className="tf-legend tf-legend--level-3"
        >
          <h4
            className="tf-legend__title"
          >
            level3
          </h4>
          <div
            className="tf-legend__list"
            style={null}
          >
            
            <div
              className="tf-legend__item tf-legend__item--square tf-legend__item--left"
            >
              <div
                className="tf-legend__symbol-container"
                style={
                  Object {
                    "width": 32,
                  }
                }
              >
                <div
                  className="tf-legend__symbol"
                >
                  <svg
                    height={16}
                    viewBox="0 0 100 100"
                    width={16}
                  >
                    <rect
                      fill="red"
                      height="100"
                      width="100"
                      x="0"
                      y="0"
                    />
                  </svg>
                </div>
              </div>
              <div
                className="tf-legend__label"
              >
                level3.1
              </div>
            </div>
            <div
              className="tf-legend__item tf-legend__item--square tf-legend__item--left"
            >
              <div
                className="tf-legend__symbol-container"
                style={
                  Object {
                    "width": 32,
                  }
                }
              >
                <div
                  className="tf-legend__symbol"
                >
                  <svg
                    height={16}
                    viewBox="0 0 100 100"
                    width={16}
                  >
                    <rect
                      fill="blue"
                      height="100"
                      width="100"
                      x="0"
                      y="0"
                    />
                  </svg>
                </div>
              </div>
              <div
                className="tf-legend__label"
              >
                level3.2
              </div>
            </div>
            <div
              className="tf-legend tf-legend--level-4"
            >
              <h4
                className="tf-legend__title"
              >
                level4
              </h4>
              <div
                className="tf-legend__list"
                style={null}
              >
                
                <div
                  className="tf-legend__item tf-legend__item--square tf-legend__item--left"
                >
                  <div
                    className="tf-legend__symbol-container"
                    style={
                      Object {
                        "width": 32,
                      }
                    }
                  >
                    <div
                      className="tf-legend__symbol"
                    >
                      <svg
                        height={16}
                        viewBox="0 0 100 100"
                        width={16}
                      >
                        <rect
                          fill="red"
                          height="100"
                          width="100"
                          x="0"
                          y="0"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    className="tf-legend__label"
                  >
                    level4.1
                  </div>
                </div>
                <div
                  className="tf-legend__item tf-legend__item--square tf-legend__item--left"
                >
                  <div
                    className="tf-legend__symbol-container"
                    style={
                      Object {
                        "width": 32,
                      }
                    }
                  >
                    <div
                      className="tf-legend__symbol"
                    >
                      <svg
                        height={16}
                        viewBox="0 0 100 100"
                        width={16}
                      >
                        <rect
                          fill="blue"
                          height="100"
                          width="100"
                          x="0"
                          y="0"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    className="tf-legend__label"
                  >
                    level4.2
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render correctly with source 1`] = `
<div
  className="tf-legend tf-legend--level-1"
>
  <h4
    className="tf-legend__title"
  >
    les bains
  </h4>
  <div
    className="tf-legend__list"
    style={null}
  >
    
  </div>
  <div
    className="tf-legend__source"
  >
    Evian
  </div>
</div>
`;
